tools/celtest/test_runner.go

```diff
@@ -890,19 +926,337 @@ func (tr *TestRunner) createResultMatcher(t *testing.T, testOutput *test.Output)
+	if details == nil || details.State() == nil {
+		return
+	}
+	if p.CoverageStats == nil {
+		p.CoverageStats = make(map[int64]set)
+	}
+	state := details.State()
+	for _, id := range state.IDs() {
+		value, found := state.Value(id)
+		if !found {
+			continue
+		}
+		if _, ok := p.CoverageStats[id]; !ok {
+			p.CoverageStats[id] = make(set)
+		}
+		p.CoverageStats[id][value] = struct{}{}
+	}
+	// Propagate visitedness after collecting direct observations
+	propagateVisitedness(p.Ast, p.CoverageStats, state)
+}
+
+// propagateVisitedness implements two rules for extending the coverage report:
+// - If a node is visited, all its ancestors are visited.
+// - If a node is visited, and it has only one child, then the child is visited.
+func propagateVisitedness(root *cel.Ast, coverageStats map[int64]set, state interpreter.EvalState) {
+	if root == nil || coverageStats == nil {
+		return
+	}
+	// Propagate visitedness upwards for ancestors.
+	ast.PostOrderVisit(root.NativeRep().Expr(), parentVisitor{coverageStats: coverageStats, evalState: state})
+	// Propagate visitedness downwards for single-child nodes.
+	ast.PostOrderVisit(root.NativeRep().Expr(), childrenVisitor{coverageStats: coverageStats, evalState: state})
+}
+
+// childrenVisitor is a visitor that populates the coverageStats for the child nodes of a given expression.
+type childrenVisitor struct {
+	coverageStats map[int64]set
+	evalState     interpreter.EvalState
+}
+
+// VisitExpr sets the coverageStats for the child of the given expression when the expression has just one child.
+func (cv childrenVisitor) VisitExpr(e ast.Expr) {
+	if cv.coverageStats[e.ID()] == nil {
+		return
+	}
+	switch e.Kind() {
+	case ast.SelectKind:
+		selExprOperand := e.AsSelect().Operand()
+		if _, ok := cv.coverageStats[selExprOperand.ID()]; !ok {
+			cv.coverageStats[selExprOperand.ID()] = make(set)
+		}
+	case ast.CallKind:
+		if c := e.AsCall(); c.FunctionName() == operators.Negate || c.FunctionName() == operators.LogicalNot {
+			if _, ok := cv.coverageStats[c.Args()[0].ID()]; !ok {
+				cv.coverageStats[c.Args()[0].ID()] = make(set)
+			}
+		}
+	}
+}
+
+// VisitEntryExpr does not affect the coverageStats for the child nodes of the given entry expression.
+func (cv childrenVisitor) VisitEntryExpr(e ast.EntryExpr) {}
+
+// parentVisitor is a visitor that populates the coverageStats of a given expression based on the
+// coverageStats of its direct children.
+type parentVisitor struct {
+	coverageStats map[int64]set
+	evalState     interpreter.EvalState
+}
+
+// VisitExpr populates the coverageStats for the current expression based on the coverageStats of its direct children.
+func (pv parentVisitor) VisitExpr(e ast.Expr) {
+	switch e.Kind() {
+	case ast.SelectKind:
+		if pv.coverageStats[e.ID()] != nil {
+			return
+		}
+		if selVal, ok := pv.coverageStats[e.AsSelect().Operand().ID()]; ok {
+			pv.coverageStats[e.ID()] = make(set)
+			pv.coverageStats[e.ID()][selVal] = struct{}{}
```

Comment:
Could you describe when this happens? 

It seems odd to associate the observed operand value to the select expression.